### PR TITLE
Remove workspace call, fix external file import via web service

### LINKED ISSUE #11718: db_import does not work when connect to the data service
## Steps to reproduce
0. Download [test.txt](https://github.com/rapid7/metasploit-framework/files/3073342/test.txt)
1. Rename `test.txt` to `test.xml` (GitHub doesn't support `.xml` file uploads)
2. Start `msfconsole`
3. Run `db_status` to confirm connection to the data service
4. `db_import path/to/test.xml`
5. Observe console output:
```
msf5 > db_import testl.xml
[*] Successfully imported path/to/test.xml
```
6. Observe no data was imported, e.g. `workspace -v`,`hosts`, `services`

## Expected behavior
```
msf5 > db_import testl.xml
[*] Importing 'Metasploit XML' data
[*] Importing host 172.16.13.2
[*] Importing host 172.16.13.1
[*] Importing host 172.16.13.254
[*] Importing host 172.16.13.144
[*] Successfully imported path/to/test.xml
```

### Metasploit version
```
msf5 > version
Framework: 5.0.17-dev-6bf824d200
Console  : 5.0.17-dev-6bf824d200
```

### I installed Metasploit with:
I have a fork of Metasploit

### OS
I observed this on MacOS

### PR DESCRIPTION
**Pull request by jrobles-r7**
The `#workspace` call wasn't running as expected when using
the data service. It was ending up in infinite recursion.

Fixes #11718. @ebleiweiss-r7 @sgonzalez-r7 

## Verification

- [x] Testing steps and file in #11718

### CODE REVIEW

**Comment by busterb** [comment]
This pattern exists in a lot of importers. They will all have the same problem too, right?

```
lib/msf/core/db_manager/import/acunetix.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/amap.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/amap.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/amap.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/appscan.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/appscan.rb:      noko_args[:wspace] = wspace
lib/msf/core/db_manager/import/burp_issue.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/burp_issue.rb:    noko_args[:wspace] = wspace
lib/msf/core/db_manager/import/burp_session.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/burp_session.rb:      noko_args[:wspace] = wspace
lib/msf/core/db_manager/import/ci.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/ci.rb:      noko_args[:wspace] = wspace
lib/msf/core/db_manager/import/foundstone.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/foundstone.rb:      noko_args[:wspace] = wspace
lib/msf/core/db_manager/import/fusion_vm.rb:    args[:wspace] ||= workspace
lib/msf/core/db_manager/import/ip360/aspl.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/ip360/v3.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/ip360/v3.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/ip_list.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/ip_list.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/libpcap.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/libpcap.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/mbsa.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/mbsa.rb:      noko_args[:wspace] = wspace
lib/msf/core/db_manager/import/metasploit_framework/credential.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/metasploit_framework/credential.rb:    wspace   = args[:wspace] || workspace
lib/msf/core/db_manager/import/metasploit_framework/xml.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/metasploit_framework/xml.rb:    wspace = args[:wspace] || workspace || Msf::Util::DBManager.process_opts_workspace(args, framework).name
lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    wspace = args[:wspace] || args['wspace'] || workspace
lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    loot_info[:workspace]  = args[:wspace]
lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    task_info[:workspace] = args[:wspace]
lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/nessus/nbe.rb:    wspace = args[:wspace] || workspace
lib/msf/core/db_manager/import/nessus/nbe.rb:    wspace = args[:wspace] || workspace
...
```

File: lib/msf/core/db_manager/import/metasploit_framework/xml.rb [hunk 1]
```diff
@@ -229,7 +229,7 @@ def import_msf_web_vuln_element(element, options={}, &notifier)
   # TODO: loot, tasks, and reports
   def import_msf_xml(args={}, &block)
     data = args[:data]
-    wspace = args[:wspace] || workspace || Msf::Util::DBManager.process_opts_workspace(args, framework).name
+    wspace = args[:wspace] || args[:workspace] || Msf::Util::DBManager.process_opts_workspace(args, framework).name
```
Comment (mkienow-r7) [inline lib/msf/core/db_manager/import/metasploit_framework/xml.rb]: The recursion was due to what `workspace` ended up calling?

Comment (mkienow-r7) [reply lib/msf/core/db_manager/import/metasploit_framework/xml.rb]: There is no reason any of the DbManager `import_*` methods should be using a pattern similar to `wspace = args[:wspace] || args[:workspace] || Msf::Util::DBManager.process_opts_workspace(args, framework).name`. It is redundant as `Msf::Util::DBManager.process_opts_workspace` calls the method `delete_opts_workspace` that pulls from `: workspace` first and then `:wspace`.

**Comment by mkienow-r7** [comment]
> This pattern exists in a lot of importers. They will all have the same problem too, right?
> 
> ```
> lib/msf/core/db_manager/import/acunetix.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/amap.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/amap.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/amap.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/appscan.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/appscan.rb:      noko_args[:wspace] = wspace
> lib/msf/core/db_manager/import/burp_issue.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/burp_issue.rb:    noko_args[:wspace] = wspace
> lib/msf/core/db_manager/import/burp_session.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/burp_session.rb:      noko_args[:wspace] = wspace
> lib/msf/core/db_manager/import/ci.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/ci.rb:      noko_args[:wspace] = wspace
> lib/msf/core/db_manager/import/foundstone.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/foundstone.rb:      noko_args[:wspace] = wspace
> lib/msf/core/db_manager/import/fusion_vm.rb:    args[:wspace] ||= workspace
> lib/msf/core/db_manager/import/ip360/aspl.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/ip360/v3.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/ip360/v3.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/ip_list.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/ip_list.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/libpcap.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/libpcap.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/mbsa.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/mbsa.rb:      noko_args[:wspace] = wspace
> lib/msf/core/db_manager/import/metasploit_framework/credential.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/metasploit_framework/credential.rb:    wspace   = args[:wspace] || workspace
> lib/msf/core/db_manager/import/metasploit_framework/xml.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/metasploit_framework/xml.rb:    wspace = args[:wspace] || workspace || Msf::Util::DBManager.process_opts_workspace(args, framework).name
> lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    wspace = args[:wspace] || args['wspace'] || workspace
> lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    loot_info[:workspace]  = args[:wspace]
> lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    task_info[:workspace] = args[:wspace]
> lib/msf/core/db_manager/import/metasploit_framework/zip.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/nessus/nbe.rb:    wspace = args[:wspace] || workspace
> lib/msf/core/db_manager/import/nessus/nbe.rb:    wspace = args[:wspace] || workspace
> ...
> ```

Yes

File: lib/msf/core/db_manager/import/metasploit_framework/xml.rb (hunk 1 above)
Comment (mkienow-r7) [reply lib/msf/core/db_manager/import/metasploit_framework/xml.rb]: If you need the workspace in DbManager methods simply use the helper as follows:
```
wspace = Msf::Util::DBManager.process_opts_workspace(opts, framework)
```

Comment (jrobles-r7) [reply lib/msf/core/db_manager/import/metasploit_framework/xml.rb]: `:workspace` argument is set here.
https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/db_manager/import.rb#L97

Recursion occurring here when using the service.
https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/db_manager/workspace.rb#L30

**Comment by jrobles-r7** [comment]
This does seem to be an issue with other importers as well.

nexpose_simplexml
```
Thin web server (v1.7.2 codename Bachmanity)
Maximum connections set to 1024
Listening on localhost:5443, CTRL+C to stop
filetype: nexpose_simplexml
#<Thread:0x00005641cc774020@/home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:18 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	11909: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:19:in `block in start_processor_thread'
	11908: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:19:in `loop'
	11907: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:22:in `block (2 levels) in start_processor_thread'
	11906: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/servlet/db_import_servlet.rb:23:in `block (2 levels) in db_import'
	11905: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/import.rb:216:in `import_file'
	11904: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/import.rb:98:in `import'
	11903: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/import/nexpose/simple.rb:16:in `import_nexpose_simplexml'
	11902: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/workspace.rb:36:in `workspace'
	 ... 11897 levels...
	    4: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/workspace.rb:36:in `workspace'
	    3: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/workspace.rb:33:in `workspace'
	    2: from /home/msfdev/git/metasploit-framework/lib/rex/logging/log_dispatcher.rb:142:in `wlog'
	    1: from /home/msfdev/git/metasploit-framework/lib/rex/logging/log_dispatcher.rb:91:in `log'
/home/msfdev/git/metasploit-framework/lib/rex/logging/log_dispatcher.rb:91:in `synchronize': stack level too deep (SystemStackError)
```

fusionvm_xml
```
Thin web server (v1.7.2 codename Bachmanity)
Maximum connections set to 1024
Listening on localhost:5443, CTRL+C to stop
filetype: fusionvm_xml
#<Thread:0x0000559e69e5a850@/home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:18 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	11909: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:19:in `block in start_processor_thread'
	11908: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:19:in `loop'
	11907: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/job_processor.rb:22:in `block (2 levels) in start_processor_thread'
	11906: from /home/msfdev/git/metasploit-framework/lib/msf/core/web_services/servlet/db_import_servlet.rb:23:in `block (2 levels) in db_import'
	11905: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/import.rb:216:in `import_file'
	11904: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/import.rb:98:in `import'
	11903: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/import/fusion_vm.rb:5:in `import_fusionvm_xml'
	11902: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/workspace.rb:36:in `workspace'
	 ... 11897 levels...
	    4: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/workspace.rb:36:in `workspace'
	    3: from /home/msfdev/git/metasploit-framework/lib/msf/core/db_manager/workspace.rb:33:in `workspace'
	    2: from /home/msfdev/git/metasploit-framework/lib/rex/logging/log_dispatcher.rb:142:in `wlog'
	    1: from /home/msfdev/git/metasploit-framework/lib/rex/logging/log_dispatcher.rb:91:in `log'
/home/msfdev/git/metasploit-framework/lib/rex/logging/log_dispatcher.rb:91:in `synchronize': stack level too deep (SystemStackError)
```

Other importers are also affected.

**Comment by mkienow-r7** [comment]
@jrobles-r7 If you didn't already catch it, there are Travis CI test failures related to the workspace.

**Comment by jrobles-r7** [comment]
All of the failures end up at the same section of code.
The `#import_msf_xml` method in `lib/msf/core/db_manager/import/metasploit_framework/xml.rb` is being called without a `:wspace` or `:workspace` argument so `Msf::Util::DBManager.process_opts_workspace(args, framework).name` is raising an exception.

Would `#import_msf_xml` be called without a wspace/workspace argument?
`#import_msf_file` calls `#import_msf_xml` without the argument but I don't see anything that calls `#import_msf_file`.

**Comment by mkienow-r7** [comment]
> All of the failures end up at the same section of code.
> The `#import_msf_xml` method in `lib/msf/core/db_manager/import/metasploit_framework/xml.rb` is being called without a `:wspace` or `:workspace` argument in the spec so `Msf::Util::DBManager.process_opts_workspace(args, framework).name` is raising an exception.
> 
> Would `#import_msf_xml` be called without a wspace/workspace argument?
> `#import_msf_file` calls `#import_msf_xml` without the argument but I don't see anything that calls `#import_msf_file`.

The method calls are dynamically created. See the `import(args={}, &block)` method in lib/msf/core/db_manager/import.rb. It should be passing `workspace` in the argument hash.
```
self.send "import_#{ftype}".to_sym, args.merge(workspace: wspace.name), &block
```

**Comment by jrobles-r7** [comment]
Yes but that's not what the spec does. Should the spec be changed or do we handle the case for a workspace not being passed into the functions.

**Comment by mkienow-r7** [comment]
Unlike most of the other `DbManager` methods these `import_#{ftype}` methods don't have a one-to-one mapping across the wire with a `DataProxy` (`lib/metasploit/framework/data_service/proxy/*`) module. The `DataProxy` ensures that a `workspace` key is added to the arguments hash. In this situation updating any spec tests to include `workspace: framework.db.workspace.name` in the arguments hash for the `import_#{ftype}` methods might be the quickest solution.

**Comment by busterb** [comment]
## Release Notes

This fixes file import functionality when using the Metasploit 5 web service.

**Pull request merged**